### DISCUSSION TITLE
Groups

### DISCUSSION THREAD

**Discussion by hifabienne** [zitadel/zitadel#2134] [Ideas]
What do you think about groups in ZITADEL? Do we need it?
What functions should groups have if we would integrate it?

**Comment by adlerhurst** [comment]
I'm thinking of two use cases (2 comments for easier threading). Both use-cases simplify authorisation management for the administrators.

In a project:

It could be useful to group roles. In this case we could authorise a user to a "role-group" instead of setting each role every time. The groups should be provided by the project managers.

**Comment by adlerhurst** [comment]
Second use case:

In an organisation:

It could be useful to group users maybe by devision (e.g. developers) and authorise a "user-group" instead of each user by itself.

**Reply by mffap** [reply]
For further reference, here the feature scope: https://github.com/zitadel/zitadel/issues/5822

**Comment by sueess** [comment]
There are two ways I can image groups in Zitadel. Manage-groups in Zitadel or groups which are exposed to the project that uses Zitadel.

* **Exposed groups**: I do not see a need for that (at least no spontaneously). I feel that roles are sufficient for that kind of grouping.
* **Manage-groups**: Groups in Zitadel could help to set permissions on role level (e. g. "who can assign a specific role to somebody"). Basically same functionality we have on organizations, which have a granted project, but in the same organization. That would allow a more detailed permission management in Zitadel.

**Comment by maennchen** [comment]
## Use Case

I'm listing two completely different use cases that I could think of using groups for. I don't think that the same group feature works for both use cases. Both are something that I could imagine calling groups.

### Climate Actions

Past Use Case we implemented via a Group Feature with Auth0:

* https://climateactions.ch
* https://joshmartin.ch/projekte/climateactions/

#### Architecture

Groups were basically used to group roles so that they could be assigned in chunks. The roles are valid globally and not only per group.

```
┌──────┐1   *┌───────────────┐*   *┌───────┐1     *┌──────┐*     1┌─────────────┐
│ User ├─────┤ Authorization ├─────┤ Group ├───────┤ Role ├───────┤ Application │
└──────┘     └───────────────┘     └───────┘       └──────┘       └─────────────┘
```

##### Roles

We defined a lot of roles for this application. Basically for every resource, we had 4 roles:

* `read:[resource]:all`
* `read:[resource]:owner`
* `write:[resource]:all`
* `write:[resource]:owner`

(And some more, but this is irrelevant for this context.)

This added up to around 120 roles for this project.

##### Groups

We had multiple groups, that allowed access via those roles. For example:

* User (default group)
  - `read:user:public`
  - `read:user:owner`
  - `write:user:owner`
  - ...
* Admin
  - `read:user:all`
  - `write:user:all`
  - ...

##### Token

This ultimately resulted in having a claim in the token / userinfo, that just listed all roles as an array.

### Hygeia

This use case is theoretical, we've already solved this using multiple organizations and granted projects.

* https://info.covid19-tracing.ch/
* https://joshmartin.ch/projekte/kanton-st-gallen-contact-tracing/
* https://github.com/jshmrtn/hygeia

In this case, having multiple organizations made sense, because the three tenants are actually separate entities. If the data belonged to a unit inside an entity (like teams in companies), I think having groups would be a better concept than setting up the IAM multiple times.

#### Architecture

We have multiple tenants in our application and we need to authorize users inside that tenant. Therefore a user is authorized to access certain roles for a group.

```
┌──────┐*      *┌───────────────┐*     1┌───────┐
│ User ├────────┤ Authorization ├───────┤ Group │
└──────┘        └───────┬───────┘       └───────┘
                        │*
                        │
                        │
                        │*
                     ┌──┴───┐
                     │ Role │
                     └──┬───┘
                        │*
                        │
                        │
                        │1
                 ┌──────┴──────┐
                 │ Application │
                 └─────────────┘
```

###### Token

To consume that structure in an application, the token / userinfo would need to contain a list of roles that mention what group they are for.

**Reply by tafli** [reply]
Use case one with groups of roles would is IMO a necessary to have a fine grained API authorization as @maennchen explained. I'm currently building an API in this manner. So I'd be glad to a have at least group of roles in Zitadel.

**Comment by Herdo** [comment]
Coming from Keycloak, groups are a feature currently missing from Zitadel. The Keycloak docs describe the use-case very well:

> Roles and groups have a similar purpose, which is to give users access and permissions to use applications. Groups are a collection of users to which you apply roles and attributes. Roles define specific applications permissions and access control.

Organizing users in logical groups and having roles assigned on both levels (users and groups) would allow easier management in larger projects and organizations with many applications.

**Reply by Herdo** [reply]
On the consumer side I'd still expect a plain list of roles the user is in, derived from the groups the user is in.

**Reply by mffap** [reply]
Hey @Herdo thanks for the input.
One way to group users within ZITADEL is with [Organization](https://docs.zitadel.com/docs/concepts/structure/organizations). This is different than Keycloak tenants, in a sense that with Projects grants you can share projects/applications across organizations. More [here](https://docs.zitadel.com/docs/guides/manage/console/organizations).
What's currently missing from your requirement would be a default way to inherit specific default roles in an organization. You could however use an Action and assign default roles to an organization.
Do you think this would work in your case?

**Reply by Herdo** [reply]
@mffap This does sound like a possibility, but seems to be a different level of grouping I'm looking for.
I'm more thinking of three dozen groups with each 5 to 50 members each.

Creating 36 organizations seems kinda overkill in the required setup and maintenance.

**Reply by mffap** [reply]
> @mffap This does sound like a possibility, but seems to be a different level of grouping I'm looking for. I'm more thinking of three dozen groups with each 5 to 50 members each.
> 
> Creating 36 organizations seems kinda overkill in the required setup and maintenance.

Maybe you could explain your use case in a bit more detail? Are the groups of users very static in terms of number of groups and users? Would you create them manually or programmatically?

We see ZITADEL being used in the way you describe. It obviously depends on what you're used to and your use case.

**Reply by Herdo** [reply]
@mffap Sure thing, I'll try to explain with a chart similiar to @maennchen . Current implementation is that users are organized in groups, and groups are assigned to roles - in Zital naming that the groups would be authorized, I guess.

```
                     ┌──────┐
                     │ User │
                     └──┬───┘
                        │*
                        │
                        │
                        │*
                    ┌───┴───┐
                    │ Group │
                    └───┬───┘
                        │*
                        │
                        │
                        │*
                ┌───────┴───────┐
                │ Authorization |
                └───────┬───────┘
                        │*
                        │
                        │
                        │*
                     ┌──┴───┐
                     │ Role │
                     └──┬───┘
                        │*
                        │
                        │
                        │1
                 ┌──────┴──────┐
                 │ Application │
                 └─────────────┘
```

The **amount of groups** would be more static, yes. With new groups coming in and old groups leaving whenever teams re-organize themself.
**Group members** would be less static, but not super volatile either. When people join or leave the company, change departmants or change jobs internally, they might need to shift to different groups.

Some example groups could be:
- `First level customer service`
- `Second level customer service`
- `Developer Team A` 
- `Developer Team B`
- `Developer Team C`
- `Developer Team D`
- `Developer Team E`
- `Operations Team 1`
- `Operations Team 2`
- `Operations Team 3`
- `Operations Team 4`
... etc

**Reply by Herdo** [reply]
Existing direct assignment `User <--> Authorization` should exist as well, but requires more management work.

**Reply by mffap** [reply]
Ok, now I understand a bit better. For your use cases, wouldn't it be feasible to group roles and apply these group of roles to users? Not clear to me why grouping users would be easier/less work in this case.

I understand however that you want to follow some organziation structure, right? But that structure is not defining the roles of an user completely, correct?

Sorry, I know a lot of questions. We get a lot of requests for 'groups' and the requirements are always a bit different 😀 Understanding what and why you need it helps us in defining a solution.

**Reply by Herdo** [reply]
@mffap Grouping roles and assigning those groups to users does only half the job.
Let's order the amount of work for each possible way of assignment I would expect:
- 🤕 **Assign roles to users directly**: High operations effort, but no upfront effort.
- 🆗 **Assign grouped roles to users directly**: Medium operations effort, low upfront effort to group roles logically.
- 🎉 **Assign grouped roles to groups of users**: Low operations effort, medium upfront effort to group roles logically and users by real-life structures.

So yes, I want to represent real-life organization structures (mostly teams, sometimes departments). This structure is not defining roles.

No worries regarding more questions 😄

**Reply by mffap** [reply]
@Herdo Thanks a lot!

@hifabienne based on this feedback and a discussion with @fforootd we might want to look into a couple of impediments with the current implementation, specifically
- inheritance of authorizations from an organization (assigning group of users roles)
- moving users between organizations (dynamic user groups)
- applying groups of roles to users and organizations (current definition of group is only for display)

**Reply by Herdo** [reply]
@mffap Just for clarification for me: so organizations will be the structure to manage groups of users, no new structure within an organization, right?

So I'll have a single organization with all our apps and roles in projects, and other organizations to organize users and grant those orgs authorizations on the "app org roles"?

**Reply by mffap** [reply]
> @mffap Just for clarification for me: so organizations will be the structure to manage groups of users, no new structure within an organization, right?
> 
> So I'll have a single organization with all our apps and roles in projects, and other organizations to organize users and grant those orgs authorizations on the "app org roles"?

For the moment I would assume that the organization structure with some more flexibility would be sufficient. Yes, although just an initial thought, but the implementation direction you describe would do it. 

A minor note: You would grant project roles and apply the authorization to users. With actions you could set those automatically when a role is granted/a new user added etc. and you would still retain the ability for self-service.

**Reply by Herdo** [reply]
Thanks! I'll do some evaluation of this solution is suitable for us. 😊

**Reply by Herdo** [reply]
@mffap After some testing this looks like a viable option.
I saw that a user who has been added to multiple organizations has the same user id displayed in the UI. Would it therefore be an options to add an existing user from another organization? If a colleague is part of multiple teams, we'd otherwise had to add the user multiple times with a details.

**Reply by Herdo** [reply]
@mffap is there a chance to see an user group concept inside an organization? So that there's no need to duplicate the users into multiple organizations?

**Reply by mffap** [reply]
One thing is that you could also give a user grant to another organization's project. With that a user should be able to access apps managed by another org.
An alternative would be to handle assignment to groups via custom metadata.
Grouping of users within an organization is not yet supported at this moment.

**Comment by T2brozz** [comment]
is there any updates on this?

**Reply by hifabienne** [reply]
Not yet, we have created one specific issue about the user group authorization here: https://github.com/zitadel/zitadel/issues/5822
If you like to see it, please upvote, so we see the demand.
At the moment we have not planned the implementation.
